### ISSUE TITLE
Add support for CANable USB to CAN interface

### ISSUE THREAD

**Issue by RayDar** [andrey-pr/OpenBafangTool#15]
I thought this would be a good place to post an an official 'TODO' for adding support for a USB to CAN adapter, such as the CANable Pro, which can be purchased online (Aliexpress) for 10 USD or so. This is the only thing keeping me (and likely many others) from using this wonderful piece of software.

**Comment by andrey-pr** [comment]
This task currently in progress 👍

**Comment by mstaack** [comment]
will something like this also be supported?

https://www.inno-maker.com/product/usb-can/

**Comment by onno204** [comment]
Hi @andrey-pr ,

I have bought a CAnable usb.
I see you have already created the branch `canable`, are there any things to keep in mind. Or can this branch be tested? (I have a Bafang m200).

**Comment by andrey-pr** [comment]
Sorry, development of canable driver is paused. Branch "canable" is currently absolutely unusable

**Comment by onno204** [comment]
Thanks for the reply.

I received my CANable-usb today, so I was keen to setting it up before seeing your comment.

I have merged the master branch into the canable branch and did _a ton_ of tinkering with the canable class.
I have finally managed to receive can data and make it show-up in the UI. 

Tomorrow I will look into sending data onto the can-bus. I hope this fixes a lot of fields being empty in the UI.

(As you can see some of the fields are still empty: https://imgur.com/a/LcdiD5D )

**Comment by onno204** [comment]
Hi @andrey-pr ,

I could use your help a little bit since I think you have already looked at this problem.
I'm having trouble collecting the serial-numbers & hw / software version.

I do receive some data on the can bus when said data is requested, but looking at the length of the data, it seems I am still missing some data?
(it is not even close to the demo data)

I do think it has something to do with the `MULTIFRAME_START = 0x04` of the CanOperationCode
Do I need to ACK the data or something?
I hope you can help me.

This is the received data:
Hardware version (3): 
`{canCommandCode: 96, canCommandSubCode: 0, canOperationCode: 4, sourceDeviceCode: 4, targetDeviceCode: 5, data: [3]}
`
Firmware version (4):
`{canCommandCode: 96, canCommandSubCode: 1, canOperationCode: 4, sourceDeviceCode: 4, targetDeviceCode: 5, data: [4]}
`
model(32):
`{canCommandCode: 96, canCommandSubCode: 2, canOperationCode: 4, sourceDeviceCode: 4, targetDeviceCode: 5, data: [32]}`

**Comment by andrey-pr** [comment]
> Hi @andrey-pr ,
> 
> I could use your help a little bit since I think you have already looked at this problem. I'm having trouble collecting the serial-numbers & hw / software version.
> 
> I do receive some data on the can bus when said data is requested, but looking at the length of the data, it seems I am still missing some data? (it is not even close to the demo data)
> 
> I do think it has something to do with the `MULTIFRAME_START = 0x04` of the CanOperationCode Do I need to ACK the data or something? I hope you can help me.
> 
> This is the received data: Hardware version (3): `{canCommandCode: 96, canCommandSubCode: 0, canOperationCode: 4, sourceDeviceCode: 4, targetDeviceCode: 5, data: [3]} ` Firmware version (4): `{canCommandCode: 96, canCommandSubCode: 1, canOperationCode: 4, sourceDeviceCode: 4, targetDeviceCode: 5, data: [4]} ` model(32): `{canCommandCode: 96, canCommandSubCode: 2, canOperationCode: 4, sourceDeviceCode: 4, targetDeviceCode: 5, data: [32]}`

This kind of parameters (version names, etc) transported several frames per parameter. One frame have limited length and this parameters are longer, so the first frame will be returned as MULTIFRAME_START, then several with MULTIFRAME, then MULTIFRAME_END and after each one you need to answer with NORMAL_ACK. Code that processes this frames is located in   [device/besst/besst.ts](https://github.com/andrey-pr/OpenBafangTool/blob/master/src/device/besst/besst.ts) file, but its not stable enough

**Comment by onno204** [comment]
Thanks for explaining the process.
I have created a generic can frame handler class, but those multiframes are all-over the place. (some targets need an ACK before sending the middle/end frame. And some targets will just send you all the multi-frames.)

I also had an issue where receiving different subcommands from the same target would just send the reply of 3 data points in a single sequence, so this required me to add a check if the message is complete at every received multiframe command.

The first mutliframe message contains the bytes' length of the incoming message. I have used this as much as possible to try and repair the received frames. When an ending frame is received, the buffer is always cleared. Broken frames are discarded, but it seems to be pretty stable now.

What seems to make the multiframes wayyy more reliable is adding a timeout between sending multiframe requests. (I have added `200ms * (SubCode % 5)` for all requests to Code 60)

I do still seem to be missing some data, but maybe that is my Bafang M200 (Veloretti) not having those data points. 
Do you have Canable device and time to test if it is the code or my engine that is causing the data to be unavailable?

Eg. when clicked the refresh button on the engine page I see 2 parameters(conParams_1 & conParams_2) not loaded:
https://imgur.com/a/bcTcryd

**Comment by willie013** [comment]
@onno204 did you manage to figure everything out for the m200 motor? And were you for instance able to set the motor power?

**Comment by bananu7** [comment]
Would it be possible to dump the CAN packets from the tool to send them "out-of-band" while the support isn't ready yet? E.g. if I wanted to use a cheap STM board with a CAN adapter, I could just hardcoded what should be sent out on the bus. I guess that can already be derived by hand from the documentation?

**Comment by andrey-pr** [comment]
> Would it be possible to dump the CAN packets from the tool to send them "out-of-band" while the support isn't ready yet? E.g. if I wanted to use a cheap STM board with a CAN adapter, I could just hardcoded what should be sent out on the bus. I guess that can already be derived by hand from the documentation?

Sorry, its not possible. Program have to request some parameters and get answer, to form correct write command

**Comment by tom-321** [comment]
What can you do with it?
Is this already supported?

https://de.aliexpress.com/item/1005005133812308.html

Does this fit this engine?

https://bafang-e.com/en/oem-area/components/component/motor/rm-g020250350500ddc

**Comment by andrey-pr** [comment]
> What can you do with it?
> Is this already supported?
> 
> https://de.aliexpress.com/item/1005005133812308.html
> 
> Does this fit this engine?
> 
> https://bafang-e.com/en/oem-area/components/component/motor/rm-g020250350500ddc

No. This device is not BESST Tool and even not a converter. Its just a hardcoded board that changes speed limit, and usb plug on it is used only for power supply

**Comment by charam73** [comment]
> > What can you do with it?
> > Is this already supported?
> > https://de.aliexpress.com/item/1005005133812308.html
> > Does this fit this engine?
> > https://bafang-e.com/en/oem-area/components/component/motor/rm-g020250350500ddc
> 
> No. This device is not BESST Tool and even not a converter. Its just a hardcoded board that changes speed limit, and usb plug on it is used only for power supply

Hi, I have this CAN/Higo (perfectly functional for my M400) gadget and it is visible in PC/Win as USB-SERIAL CH340 and in your app as UART/Com device, but no data read from CANbus..

**Comment by minax007** [comment]
Hi @andrey-pr 
I have just found this repository which might be useful for you:
https://github.com/bart594/bafang_canable_pro

**Comment by due2expire2** [comment]
> > > What can you do with it?
> > > Is this already supported?
> > > https://de.aliexpress.com/item/1005005133812308.html
> > > Does this fit this engine?
> > > https://bafang-e.com/en/oem-area/components/component/motor/rm-g020250350500ddc
> > 
> > 
> > No. This device is not BESST Tool and even not a converter. Its just a hardcoded board that changes speed limit, and usb plug on it is used only for power supply
> 
> Hi, I have this CAN/Higo (perfectly functional for my M400) gadget and it is visible in PC/Win as USB-SERIAL CH340 and in your app as UART/Com device, but no data read from CANbus..

I would like to second this. I bought a CAN programmer advertised for CAN Versions of the BBS02 from Varstrom. It also shows up as a USB Serial CH340, but does not register in the tool as a CAN device. The software that came with the programmer can change *some* variables, so it does work.

**Comment by minax007** [comment]
Meanwhile there is a fork from the repository that I had posted four months ago which is well documented and provides binaries of the server by which you can easily access the GUI via a web browser:

https://github.com/mdi-9/bafang_canable_pro

The adapter to use this software costs 4 EUR and can be used out of the box without any adaptions needed.
